### PR TITLE
[WIP] Document guidelines for saving or submitting data

### PR DESCRIPTION
**Pull request by mperrotti**
## TODO
- [x] Replace current screenshot examples with mocks that are more "zoomed in" and specific to what is being demonstrated
- [x] Try to simplify the button placement guidelines
- [x] Page formatting/layout
- [ ] Try a version that doesn't mention controls that don't exist yet (segmented control, toggle switch)
- [ ] Confirm that all the questions under the "Questions these docs aim to answer" heading are answered
- [ ] Proofread

---

## Use cases I looked at when developing these guidelines
- Settings forms (repo, user)
- Editing an issue from the Memex table
- Creating a PR review
- Changing assignees/labels/milestones/etc on an issue
- Creating a repo
- Creating a gist
- Configuring a GitHub Action

## Questions these docs aim to answer
- When is it ok to save a change without explicitly clicking a CTA? When is that ok and when isn't it ok?
- What do I do with a save CTA button when no data has been changed?
- How do I keep the user informed that the form submission is still in progress?
- How do I prevent changes or multiple submissions while an entry is bing submitted?
- Can I save a form in segments? When is it ok or not ok to do?
- How do I decide what happens after I've saved?
- When do we need to give feedback that something has been saved successfully or that there was an error saving?
    - How should we show that feedback? 
    - How does the way we show that feedback change depending on whether there is a page refresh required? For example: We can't show something fleeting (like a toast) if the page requires a refresh to save
- Where do "Save" buttons get placed?
    - When it's saving an inline edit?
    - When it's for a small form within a dialog?
    - When it's for a full-page form?
- How do we make sure people don't accidentally lose data?
- What considerations do we make for keyboard-only users?
    - Should all forms be submit-able using the "enter" key?
    - Should a multiselect dropdown's values be canceled if I hit the "Esc" key?
- Where should focus go after I save/submit?
    - Keep it on the button unless:
        - The button is no longer there (e.g.: 1. leaving a comment and the form disappears, 2. after submission I get navigated to a new page)
        - The button becomes disabled
        - There is an important message (like an error message) that the user needs to see. Then, focus the message
- How do I handle saving a form inside of a dropdown?
    - A11y team is recommending we don't automatically save/submit forms in a dropdown - the user must explicitly submit the form with a button
    - What are the cases where we don't require the user to explicitly submit the form with a button?

## What is out-of-scope for now
- Autosaving documents
- Previewing the effects of the changes
- Saved drafts

---

## Related
- https://github.com/github/accessibility/issues/415
- [Google Doc with content revisions](https://docs.google.com/document/d/1-l-_3bmJMqNTqUXUear3rhSRnkkgVdtw6cdhkbDJoNw/edit#)
- https://github.com/github/design-systems/pull/309 includes examples ([new link to the .md with user stories](https://github.com/github/design-systems/blob/master/archive/interaction-guidelines/user-stories.md))
- 🔍 [Figma audit file](https://www.figma.com/file/ldm5ZGxRgrQ1KCiDY7qvdh/Saving-models?node-id=1%3A3)

### CODE REVIEW

**Comment by mperrotti** [comment]
Things I'm still unsure about and want to ask the design team about:
- Should we consider a sticky header for form actions (e.g.: save, cancel)?
- Should save/cancel buttons always be on the bottom right?
- Save/cancel at the top _and_ bottom of a form if it's a long form?
- Should it ever be acceptable to show the browser-native alert that warns users they might lose unsaved data when they navigate away from the page?

File: content/ui-patterns/saving.mdx [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+- use a sticky header with save and cancel buttons?
+*/}
+
+If a save button is used to save an inline edit to the input or inputs where the change was made.
+![Screen recording of a user editing a PR's title](https://user-images.githubusercontent.com/2313998/150868914-6af36067-39a6-4fec-adb9-518c5abd595d.gif)
+
+It's strongly recommended to have a single save button for all of the inputs on the page. If you're designing a page with pieces of content that can be separately toggled into an edit mode (such as editing an issue title), only allow one edit mode toggled on at a time so that only one save button is visible. When there is on save button per section of a page or form, users may be confused about whether the save button will save all of the inputs or just the subset of the inputs. This confusion is even more likely for visually impaired users because they can't see if each save button is visually grouped with a set of inputs.
+
+{/* 
+TODO: find an example of this, mock one up (if that's allowed), or exclude this visual example
```
Comment (mperrotti) [inline content/ui-patterns/saving.mdx:39]: Any ideas for _good_ examples of a form with subsection that gets submitted with a "Save" button, and a separate subsection that automatically saves?

The closest thing I could find was the Codespaces settings - you have to explicitly click "save" for one section, but the rest automatically save.

File: content/ui-patterns/saving.mdx [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+If the change may not be obvious in the UI within the user's viewport, provide feedback to let the user know that the change occurred successfully. If the change occured without a redirect or page refresh, you could use a toast message. If the change occured and the page refreshes or redirects, you could use a flash banner message.
+
+<img width="1787" alt="Screenshot of the user's activity feed page with a flash banner confirming that a repository has been deleted" src="https://user-images.githubusercontent.com/2313998/150868934-b96e6d9d-d022-433f-87ab-189baf1776fa.png">
+
+
+If an imperitive input did not automatically save, show an error message that explains why the change wasn't saved.
+
+{/*
+TODO: find an example of this, mock one up (if that's allowed), or exclude this visual example
```
Comment (mperrotti) [inline content/ui-patterns/saving.mdx:123]: Any ideas where I could find an auto-saving field that shows an error message when the change encounters an error?

**Review by mperrotti — COMMENTED** [review]
@yaili - if I can't find any good examples to screenshot, is it acceptable to mock up a fake example?

File: content/ui-patterns/saving.mdx [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+- Sensitive information could be unintentionally submitted. For example: a change password input
+
+**Checkbox groups**
+- It could be unclear whether or not their selections have been saved
+- Users may accidentally submit a selection that they clicked by accident
+
+**Radio button groups**
+Screenreader users can't read the options without selecting them. This could accidentally apply a selection the user didn't want.
+
+**Browser-native `<select>` dropdowns**
```
Comment (ichelsea) [inline content/ui-patterns/saving.mdx:109]: This also happens on macs too. We should clarify in our implicit and explicit input examples.

File: content/ui-patterns/saving.mdx [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+On Windows, a select dropdown has a similar problem as a radio button group. 
+Options are selected when a user focuses the select and uses their arrow keys to read through each option
+
+## Feedback
+
+If the change may not be obvious in the UI within the user's viewport, provide feedback to let the user know that the change occurred successfully. If the change occured without a redirect or page refresh, you could use a toast message. If the change occured and the page refreshes or redirects, you could use a flash banner message.
```
Comment (ichelsea) [inline content/ui-patterns/saving.mdx:115]: We might have to not recommend toast messages until we fix the accessibility issues.

File: content/ui-patterns/saving.mdx [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+<img width="1787" alt="Screenshot of the user's activity feed page with a flash banner confirming that a repository has been deleted" src="https://user-images.githubusercontent.com/2313998/150868934-b96e6d9d-d022-433f-87ab-189baf1776fa.png">
```
Comment (ichelsea) [inline content/ui-patterns/saving.mdx:117]:
```suggestion
<img width="1787" alt="Flash banner on activity feed page confirming a successful repository deletion" src="https://user-images.githubusercontent.com/2313998/150868934-b96e6d9d-d022-433f-87ab-189baf1776fa.png">
```

File: content/ui-patterns/saving.mdx [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+TODO: find an example of this, mock one up (if that's allowed), or exclude this visual example
+
+![🖼 Include a DO example with image to reinforce message]()
+*/}
+
+## Error forgiveness
+
+For risky changes, provide an option to "undo" after saving.
+
+![Screenshot of a user hovering the 'Revert' button after merging a PR](https://user-images.githubusercontent.com/2313998/150870681-349185eb-342a-49f2-a7d3-6e58132e86a0.png)
```
Comment (ichelsea) [inline content/ui-patterns/saving.mdx:132]:
```suggestion
![Mouse cursor hovering over the 'Revert' button after merging a PR](https://user-images.githubusercontent.com/2313998/150870681-349185eb-342a-49f2-a7d3-6e58132e86a0.png)
```

File: content/ui-patterns/saving.mdx [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+
+Before submitting a form with very dangerous consequences (for example, delete a repository), ask the to user confirm they want to submit the form.
+
+![Screen recording of a user confirming that they want to delete a repository](https://user-images.githubusercontent.com/2313998/150868926-f99d23f6-6e31-4333-9545-2625451dae99.gif)
```
Comment (ichelsea) [inline content/ui-patterns/saving.mdx:137]:
```suggestion
![User flow of a user confirming that they want to delete a repository](https://user-images.githubusercontent.com/2313998/150868926-f99d23f6-6e31-4333-9545-2625451dae99.gif)
```

**Review by ichelsea — COMMENTED** [review]
This is looking great! Thanks for being the one to start pulling this guidance together 🙂

**Comment by yaili** [comment]
> @yaili - if I can't find any good examples to screenshot, is it acceptable to mock up a fake example?

Yes, I think that's probably ideal even: to show an ideal but realistic example that is easier to parse than just a screenshot.

File: content/ui-patterns/saving.mdx [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+<img width="1790" alt="Screenshot of the 'Edit repository details' form" src="https://user-images.githubusercontent.com/2313998/150868909-71f72f1a-eafc-4777-a639-96d81b33eebb.png">
+
+{/*
+TODO:
+Consider how we might define more patterns to keep the "Save" button easily accessible on long forms:
+- add save and cancel buttons on top AND bottom?
+- use a sticky header with save and cancel buttons?
+*/}
+
+If a save button is used to save an inline edit to the input or inputs where the change was made.
```
[the remark this answers is not in the document]
Comment (mperrotti) [reply content/ui-patterns/saving.mdx:33]: This was a typo. It should be something like:

> If a save button is used to save an inline edit, place it close to the input where the change was made.

---

That wording could still use a little help, so I'll spend some more time on it.

File: content/ui-patterns/saving.mdx [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+TODO: find an example of this, mock one up (if that's allowed), or exclude this visual example
+
+![🖼 Include a DO example with image to reinforce message]()
+*/}
+
+{/* TODO: reference the case where there's a declaratively saved form on a page with inherently saved inputs (such as user Profile settings) */}
+
+### State
+
+Do not disable or hide a form's save button if the form is invalid or has not been changed. The only time the save button should be disabled is in cases where you need to prevent the user from taking further action while the request completes (such as updating notification preferences). In these cases, disable the save button and the form inputs. While the form is disabled, update the save button's label with the present continuous form of the action's verb (for example, "Save" becomes "Saving...", "Follow" becomes "Following...") to indicate that they shouldn't navigate away from the current page which may cause the request to fail silently in the event of an error. In most cases, these actions should take minimal time, but in cases where it takes longer than normal, this clear indication is invaluable.
```
[the remark this answers is not in the document]
Comment (mperrotti) [reply content/ui-patterns/saving.mdx:48]: Thanks for the context. I'll hold off on this for now.

File: content/ui-patterns/saving.mdx [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+
+![Screenshot of a CTA label that changes from "Update" to "Updating..."](https://user-images.githubusercontent.com/2313998/150868916-ea209d7c-136e-4f6a-a1b0-51d1e54faf46.png)
+
+## Explicit saving
+An explicit save does not apply changes until the user explicitly takes an action to submit the changes.
+
+If you're designing a form, start with an explicit saving pattern. Avoid mixing paradigms on a single page with multiple forms, and never mix paradigms in a single form. For example: if I upload a profile photo in the "Public profile" form, that change should not be saved until I explicitly submit the whole form.
```
[the remark this answers is not in the document]
Comment (mperrotti) [reply content/ui-patterns/saving.mdx:55]: Good call - "patterns" is simpler and clearer. I also think that "**implicit** save patterns" could be phrased more simply. In the documentation, I refer to implicit save as "automatic save".

File: content/ui-patterns/saving.mdx [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+If a user tries to navigate away from a page with an unsaved form, we should warn them that their changes will be lost.
+*/}
+
+### Declarative inputs
+
+When designing a form that is saved explicitly, use the following inputs:
+- Text inputs
+- Checkbox groups
+- Radio button groups
+- Dropdown menus
```
[the remark this answers is not in the document]
Comment (mperrotti) [reply content/ui-patterns/saving.mdx:71]: Yea, we could be more specific here.

Instead of listing all the possible kinds of dropdowns ([ActionMenu](https://primer.style/react/storybook?path=/story/composite-components-actionmenu2--simple-list-story), [SelectPanel](https://primer.style/react/storybook?path=/story/composite-components-selectpanel--multi-select-story), [DropdownMenu](https://primer.style/react/storybook?path=/story/composite-components-dropdownmenu2--simple-dropdown-menu)), this summary could cover everything:
```suggestion
- Browser-native `<select>` dropdown menus
- Custom dropdown menus that allow multiple selections
```

File: content/ui-patterns/saving.mdx [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+
+## Automatic saving
+
+Automatic saving applies changes as they are made.
+
+Automatic saving should be used when the user expects instant feedback from the change they made. Changes should only be saved automatically using[imperitive inputs](#imperitive-inputs) and for individual settings that are not changed via separate flows.
+
+If a field is saved automatically, it should be obvious whether or not it saved without using text or visual indicators. A visual indicator for a field that has been "saved" or has an error "error" could be mistaken as a validation status.
+
+![Screen recording of a user changing their theme in the 'Appearance' settings tab](https://user-images.githubusercontent.com/2313998/150868921-afc2b8b8-e39f-478f-b17d-398b790bda89.gif)
```
[the remark this answers is not in the document]
Comment (mperrotti) [reply content/ui-patterns/saving.mdx:83]: Yea, the screenshot examples are problematic. Instead of doing screenshots, I'm going to mock up very targeted examples.

File: content/ui-patterns/saving.mdx [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+
+
+### Imperitive inputs
+
+When user input is automatically saved, use the following inputs:
+- Toggle switches
+- Segmented controls
+- Single-select dropdowns
```
[the remark this answers is not in the document]
Comment (mperrotti) [reply content/ui-patterns/saving.mdx:91]: > If you've got a legit single select menu, like the watch menu, then that wouldn't need a save button

This is the case I was talking about, but I see how that's not clear here.

You were spot-on with your example. We wouldn't want to automatically save a dropdown selection if it's part of a form with a save button.

File: content/ui-patterns/saving.mdx [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+
+These inputs are designed to reduce UI by giving instant feedback. The toggle switch behaves like a light switch: the light just switches on and off without you having to confirm each time you flip the switch. Segmented controls and single-select dropdowns behave like a soda machine: the soda comes out as soon as you press the buttons without you confirming which soda you wanted.
+
+To build an accessible and consistent app, do not automatically save declarative inputs.
+
+### Inputs that are problematic for autosaving:
+**Text inputs**
```
[the remark this answers is not in the document]
Comment (mperrotti) [reply content/ui-patterns/saving.mdx:98]: Yes, absolutely.

File: content/ui-patterns/saving.mdx (hunk 4 above)
Comment (mperrotti) [reply content/ui-patterns/saving.mdx:115]: Are the accessibility issues caused by the frontend implementation? Or are we designing with toasts in a way that is not accessible?

If it's just about the frontend implementation, we could keep the guidance but add a note that toasts should be avoided until we fix them.

File: content/ui-patterns/saving.mdx (hunk 1 above)
Comment (mperrotti) [reply content/ui-patterns/saving.mdx:39]: Ignore this comment. I'm going to be replacing all screenshots with mockups that demonstrate the issue in isolation.

File: content/ui-patterns/saving.mdx (hunk 2 above)
Comment (mperrotti) [reply content/ui-patterns/saving.mdx:123]: Ignore this comment. I'm going to be replacing all screenshots with mockups that demonstrate the issue in isolation.

**Comment by mperrotti** [comment]
Note to reviewers: Please disregard the screenshot examples. I'm going to be replacing all screenshots with mockups that demonstrate the pattern in isolation.

File: content/ui-patterns/saving.mdx (hunk 12 above)
[the remark this answers is not in the document]
Comment (ichelsea) [reply content/ui-patterns/saving.mdx:83]: Let me know if you want help to mock those up

File: content/ui-patterns/saving.mdx (hunk 14 above)
[the remark this answers is not in the document]
Comment (ichelsea) [reply content/ui-patterns/saving.mdx:98]: Yes I believe so

File: content/ui-patterns/saving.mdx [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+{/*
+TODO:
+1. Be more specific about when a button should be placed somewhere besides the bottom-right. Define alternative patterns and when to use them.
+
+2. Consider how we might define more patterns to keep the "Save" button easily accessible on long forms:
+- add save and cancel buttons on top AND bottom?
+- use a sticky header with save and cancel buttons?
+*/}
+
+To minimize UI, some content may be edited inline without taking the user to a separate flow. An input used for inline editing should have a save button placed very close to visually group the input with the button.
```
Comment (yaili) [inline content/ui-patterns/saving.mdx:55]:
```suggestion
To simplify the interface, some content may be edited inline without taking the user to a separate flow. An input used for inline editing should have a save button placed very close to visually connect the input with the button.
```

Minimize could be confuse with "minimize a window".

File: content/ui-patterns/saving.mdx [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+
+![Save button is placed to the right of the input used to edit the PR title](https://user-images.githubusercontent.com/2313998/150868914-6af36067-39a6-4fec-adb9-518c5abd595d.gif)
+
+If you encounter a case where your submit button would not be intuitive or easily accessible using any of the placement options described above, then consider alternatives and optinally propose a new Primer pattern.
```
Comment (yaili) [inline content/ui-patterns/saving.mdx:59]:
```suggestion
If you believe your submit button will not be intuitive or easily accessible using any of the placement options listed here, you may consider alternatives and optionally propose a new Primer pattern.
```

File: content/ui-patterns/saving.mdx [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+
+### State
```
Comment (yaili) [inline content/ui-patterns/saving.mdx:61]: Because state is so small, I'm wondering if you should move it to before placement.

File: content/ui-patterns/saving.mdx [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+Screenreader users can't read the options without selecting them. This could accidentally apply a selection the user didn't want.
+
+**Browser-native `<select>` dropdowns**
+A select dropdown has a similar problem as a radio button group. Options are selected when a user focuses the select and uses their arrow keys to read through each option
+
+## Automatic saving
+
+Automatic saving applies changes as they are made.
+
+Automatic saving should be used when the user expects instant feedback from the change they made. Changes should only be saved automatically using [imperitive inputs](#imperitive-inputs) and for individual settings that are not changed via separate flows.
```
Comment (yaili) [inline content/ui-patterns/saving.mdx:108]: Do you mean imperative? If so, find a replace in the entire doc 😬

Comment (yaili) [reply content/ui-patterns/saving.mdx:108]: > individual settings that are not changed via separate flows.

I don't understand what this means.

File: content/ui-patterns/saving.mdx [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+
+If a field is saved automatically, it should be obvious whether or not it saved without using text or visual indicators. A visual indicator for a field that has been "saved" or has an error "error" could be mistaken as a validation status.
```
Comment (yaili) [inline content/ui-patterns/saving.mdx:110]:
```suggestion
If a form field is saved automatically, it should be obvious whether or not it saved without using text or visual indicators. A visual indicator for a form field that has been "saved" or has an error "error" could be mistaken as a validation status.
```

Comment (yaili) [reply content/ui-patterns/saving.mdx:110]: > A visual indicator for a field that has been "saved" or has an error "error" could be mistaken as a validation status.

I'm finding this sentence really hard to parse. Can you reword?

File: content/ui-patterns/saving.mdx [hunk 20]
```diff
@@ -0,0 +1,165 @@
...
+
+![Screen recording of a user changing their theme in the 'Appearance' settings tab](https://user-images.githubusercontent.com/2313998/150868921-afc2b8b8-e39f-478f-b17d-398b790bda89.gif)
+
+
+### Imperitive inputs
+
+When user input is automatically saved, use the following inputs:
```
Comment (yaili) [inline content/ui-patterns/saving.mdx:117]:
```suggestion
Imperative inputs are designed to reduce UI by giving instant feedback. When user input is automatically saved, use the following inputs:
```

File: content/ui-patterns/saving.mdx [hunk 21]
```diff
@@ -0,0 +1,165 @@
...
+- Toggle switches
+- Segmented controls
+- Single-select dropdowns that are not native `<select>` dropdowns or part of an explicitly saved form
+
+These inputs are designed to reduce UI by giving instant feedback. The toggle switch behaves like a light switch: the light just switches on and off without you having to confirm each time you flip the switch. Segmented controls and single-select dropdowns behave like a soda machine: the soda comes out as soon as you press the buttons without you confirming which soda you wanted.
```
Comment (yaili) [inline content/ui-patterns/saving.mdx:122]:
```suggestion
The toggle switch behaves like a light switch: the light just switches on and off without you having to confirm each time you flip the switch. Segmented controls and single-select dropdowns behave like a soda machine: the soda comes out as soon as you press the button to select which soda you want.
```

File: content/ui-patterns/saving.mdx [hunk 22]
```diff
@@ -0,0 +1,165 @@
...
+![Mouse cursor hovering over the 'Revert' button after merging a PR](https://user-images.githubusercontent.com/2313998/150870681-349185eb-342a-49f2-a7d3-6e58132e86a0.png)
+
+
+Before submitting a form with very dangerous consequences (for example, delete a repository), ask the to user confirm they want to submit the form.
+
+![User flow of a user confirming that they want to delete a repository](https://user-images.githubusercontent.com/2313998/150868926-f99d23f6-6e31-4333-9545-2625451dae99.gif)
+
+{/*TODO: come up with some helpful guidance on tabbed forms, or just exclude these guidelines until an issue comes up*/
+
+## Saving multi-page forms
```
Comment (yaili) [inline content/ui-patterns/saving.mdx:150]: Please a little intro sentence here :)

File: content/ui-patterns/saving.mdx [hunk 23]
```diff
@@ -0,0 +1,165 @@
...
+- Checkbox groups
+- Radio button groups
+- Browser-native `<select>` dropdown menus
+- Custom dropdown menus that allow multiple selections
+
+To maintain consistent interaction patterns across your app, do not use toggle switches or segmented controls in an explicitly saved form. They should only be used when they automatically change based on user input.
+
+To build an accessible and consistent app, do not automatically save declarative inputs.
+
+### Problems with autosaving:
```
Comment (yaili) [inline content/ui-patterns/saving.mdx:89]: I feel that this section should be in the automatic saving section. I know it's related to why avoid these components when explicit saving, but it feels weird. 🤔

Comment (yaili) [reply content/ui-patterns/saving.mdx:89]: I'll need to do another review of this section once it's moved btw.

File: content/ui-patterns/saving.mdx [hunk 24]
```diff
@@ -0,0 +1,165 @@
...
+When user input is automatically saved, use the following inputs:
+- Toggle switches
+- Segmented controls
+- Single-select dropdowns that are not native `<select>` dropdowns or part of an explicitly saved form
+
+These inputs are designed to reduce UI by giving instant feedback. The toggle switch behaves like a light switch: the light just switches on and off without you having to confirm each time you flip the switch. Segmented controls and single-select dropdowns behave like a soda machine: the soda comes out as soon as you press the buttons without you confirming which soda you wanted.
+
+## Feedback
+
+If the change may not be obvious in the UI within the user's viewport, provide feedback to let the user know that the change occurred successfully. If the change occured without a redirect or page refresh, you could use a toast message. If the change occured and the page refreshes or redirects, you could use a flash banner message.
```
Comment (yaili) [inline content/ui-patterns/saving.mdx:126]:
```suggestion
If the change is not obvious within the user's viewport, provide feedback to let the user know that the change occurred successfully. If the change occurred without a redirect or page refresh, you can use the toast component. If the change occurred and the page refreshes or redirects, you can use a flash banner component.
```

File: content/ui-patterns/saving.mdx [hunk 25]
```diff
@@ -0,0 +1,165 @@
...
+
+**A note on toasts:** If your design is being implemented in github.com, avoid using a toast for now. They have accessibility issues that need to be fixed.
```
Comment (yaili) [inline content/ui-patterns/saving.mdx:128]: This should use the note component: https://primer.style/doctocat/components/note

File: content/ui-patterns/saving.mdx [hunk 26]
```diff
@@ -0,0 +1,165 @@
...
+## Explicit saving
+An explicit save does not apply changes until the user explicitly takes an action to submit the changes.
+
+Explicitly saved forms can be saved by clicking a submit button or by pressing the "Enter" key while focused on an input.
+
+If you're designing a form, start with an explicit saving pattern. Avoid mixing explicit and automatic save patterns on a single page with multiple forms, and never mix save patterns in a single form. For example: if I upload a profile photo in the "Public profile" form, that change should not be saved until I explicitly submit the whole form.
+
+If there was an error saving the data, the user's data should be preserved in the form and they should be given [feedback](#feedback) about the failure.
+
+If a user tries to navigate away from a page with a form that has unsaved changes, you can optionally warn them that their changes will be lost using the [beforeunload event](https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event). Modern browsers prevent developers from customizing the alrt message, so we're limited to a generic message like "Do you want to leave this site? Changes you made may not be saved."
```
[the remark this answers is not in the document]
Comment (mperrotti) [reply content/ui-patterns/saving.mdx:139]: No, that was a typo. I fixed it when implementing your suggestion.

File: content/ui-patterns/saving.mdx (hunk 18 above)
Comment (mperrotti) [reply content/ui-patterns/saving.mdx:183]: It means edits that don't bring you into a new page or modal dialog. For example: adding a milestone to an issue.

I'm just removing that part because I think the first part of the sentence covers it.

**Comment by mperrotti** [comment]
Review this PR instead: https://github.com/primer/design/pull/212
